### PR TITLE
convert elfobj.c to D

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by WalterBright** [comment]
Apparently `build.d` is not compiling in `elfobj.d`. Anyone got a quick fix?

**Comment by rainers** [comment]
> Apparently `build.d` is not compiling in `elfobj.d`. Anyone got a quick fix?

It seems to be filtered by this line: https://github.com/dlang/dmd/blob/master/src/build.d#L802 and added as a "header" a few lines below.

**Comment by WalterBright** [comment]
`canFind("dt.d", "obj.d")` shouldn't match `elfobj.d`

**Comment by rainers** [comment]
> `canFind("dt.d", "obj.d")` shouldn't match `elfobj.d`

? I do see the needle "obj.d" in the haystack "elfobj.d": https://dlang.org/phobos/std_algorithm_searching.html#.canFind. subranges are also considered "needles".

**Comment by WalterBright** [comment]
Look lower a bit at the "beeswax" example. "bees" doesn't match.

**Comment by rainers** [comment]
> Look lower a bit at the "beeswax" example. "bees" doesn't match.

https://run.dlang.io/is/c4pLng

**Comment by WalterBright** [comment]
So some `canFind()`s return a `bool` and some return a `size_t`. Grrrrr.

"So, if used directly in the condition of an if statement or loop, the result will be true if one of the needles is found and false if none are found, whereas if the result is used elsewhere, it can either be cast to bool for the same effect or used to get which needle was found first without having to deal with the tuple that LREF find returns for the same operation."

Gobbledygook, while evidently having quite different behavior from the beeswax one.

**Comment by rainers** [comment]
The beeswax example is different because it operates on string[], not string.

**Comment by WalterBright** [comment]
You're right. But I found a workaround anyway. Thanks for pointing me to the right place. BTW, the design of build.d relying on reading the directory and blacklisting instead of working off of a whitelist is not so robust, as this problem illustrates.

File: src/build.d [hunk 1]
```diff
@@ -800,7 +800,7 @@ auto sourceFiles()
             dirEntries(env["C"], "*.d", SpanMode.shallow)
                 .map!(e => e.name)
                 .filter!(e => !e.canFind("dt.d", "obj.d"))
```
Comment (wilzbach) [inline src/build.d]: Why not `buildPath(env["C"], "obj.d")`

**Comment by WalterBright** [comment]
What happened to the circleci builds?

**Comment by thewilsonator** [comment]
I _believe_ that Circle was running the user project test which are now handled by build kite.

File: src/build.d (hunk 1 above)
Comment (WalterBright) [reply src/build.d]: ok

**Comment by thewilsonator** [comment]
Oh wait, they appeared now.

File: src/dmd/backend/elfobj.d [hunk 2]
```diff
@@ -260,8 +249,8 @@ private IDXSEC secidx_note;      // Final table index for note data
 
 // Comment data for compiler version
 private Outbuffer *comment_data;
-private const char compiler[] = "\0Digital Mars C/C++"
-        VERSION
+private const(char)* compiler = "\0Digital Mars C/C++" ~
+        " 2.083" //VERSION
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d]: `__VERSION__.stringof`?

File: src/dmd/backend/elfobj.d [hunk 3]
```diff
@@ -291,22 +280,22 @@ enum
...
+Elf32_Shdr* MAP_SEG2SEC(int seg)    { return &SecHdrTab[MAP_SEG2SECIDX(seg)]; }
+int         MAP_SEG2TYP(int seg)    { return MAP_SEG2SEC(seg).sh_flags & SHF_EXECINSTR ? CODE : DATA; }
 
-seg_data **SegData;
-int seg_count;
+public seg_data **SegData;
+public int seg_count;
 int seg_max;
 int seg_tlsseg = UNKNOWN;
 int seg_tlsseg_bss = UNKNOWN;
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d]: `__gshared`

File: src/dmd/backend/elfobj.d [hunk 4]
```diff
@@ -1909,25 +1916,25 @@ int Obj_jmpTableSegment(Symbol *s)
...
     {
         seg_data *pgroupseg = SegData[groupseg];
 
         /* Don't write it if it is already there
          */
-        Outbuffer *buf = pgroupseg->SDbuf;
-        assert(sizeof(int) == 4);               // loop depends on this
-        for (size_t i = buf->size(); i > 4;)
+        Outbuffer *buf = pgroupseg.SDbuf;
+        assert(int.sizeof == 4);               // loop depends on this
```
Comment (thewilsonator) [inline src/dmd/backend/elfobj.d]: `static assert`? Or just remove entirely. Fine if done later.

File: src/build.d (hunk 1 above)
Comment (wilzbach) [reply src/build.d]: Oh I was referring to the filter line, because then you wouldn't need ty readd it at the end, no?

**Review by thewilsonator — CHANGES_REQUESTED** [review]

**Review by thewilsonator — APPROVED** [review]

**Comment by WalterBright** [comment]
Blocking #8911

**Comment by ibuclaw** [comment]
This introduced _another_ regression.

https://semaphoreci.com/dlang/dmd-2/branches/master/builds/1211

Now LDC builds are failing.

**Pull request merged**